Laboratory/Tests/L4-Standard-Library/Matrix2x2.cs

```diff
@@ -0,0 +1,134 @@
+﻿using NUnit.Framework;
+
+namespace Laboratory.Tests.L4.StandardLibrary
+{
+    internal class Matrix2x2 : StandardLibraryFixture
+    {
+
+        public static (string constructorArgs, string constant)[] ConstantArgList =
+        {
+            ("(Vector2(1, 0), Vector2(0, 1))", "Matrix2x2.identity"),
+        };
+        [Test]
+        public void Constants([ValueSource(nameof(ConstantArgList))] (string lhs, string rhs) args, [Values] EvaluationMode mode)
+        {
+            AssertApproxEqual(ValidatedCompilerInput,
+                              new FunctionEvaluation("Matrix2x2", args.lhs, mode),
+                              new ExpressionEvaluation(args.rhs, mode));
+        }
+
+        public static (string constructorArgs, string factory, string factoryArgs)[] FactoryArgsList =
+        {
+            ("(Vector2(1, 2), Vector2(3, 4))", "fromRows", "(Vector2(1, 2), Vector2(3, 4))"),
+            ("(Vector2(1, 2), Vector2(3, 4))", "fromCols", "(Vector2(1, 3), Vector2(2, 4))"),
+            ("(Vector2(3, 0), Vector2(0, 5))", "fromDiagonal", "(Vector2(3, 5))"),
+        };
+        [Test]
+        public void Factory([ValueSource(nameof(FactoryArgsList))] (string lhs, string factory, string rhs) args, [Values] EvaluationMode mode)
+        {
+            string defaultConstructor = "Matrix2x2";
+            string altConstructor = "Matrix2x2." + args.factory;
+            AssertApproxEqual(ValidatedCompilerInput,
+                              new FunctionEvaluation(defaultConstructor, args.lhs, mode),
+                              new FunctionEvaluation(altConstructor, args.rhs, mode));
+        }
+
+        public static (string constructorArgs, string property, string result)[] GetPropertiesArgsList =
+        {
+            ("(Vector2(1, 2), Vector2(3, 4))", "xCol", "Vector2(1, 3)"),
+            ("(Vector2(1, 2), Vector2(3, 4))", "yCol", "Vector2(2, 4)"),
+            ("(Vector2(1, 2), Vector2(3, 4))", "xRow", "Vector2(1, 2)"),
+            ("(Vector2(1, 2), Vector2(3, 4))", "yRow", "Vector2(3, 4)"),
+            ("(Vector2(1, 2), Vector2(3, 4))", "diagonal", "Vector2(1, 4)"),
+        };
+        [Test]
+        public void GetProperties([ValueSource(nameof(GetPropertiesArgsList))] (string lhs, string prop, string rhs) args, [Values] EvaluationMode mode)
+        {
+            // Tests which construct the matrix and get the named property
+
+            string getter = "_(v1:Vector2, v2:Vector2) = Matrix2x2(v1, v2)." + args.prop;
```

Comment:
I think you'll need to add return types to these lambdas so they work at  runtime (compiled)

Reply:
Done 👍

Reply:
sweet :)